src/storage-controller/src/history.rs

```diff
@@ -155,7 +160,16 @@ impl<T: std::fmt::Debug> CommandHistory<T> {
                 if frontier.is_empty() {
                     continue;
                 }
-                sink.description.as_of = frontier;
+                // The as-of is at least the implied capability of the sink collection. If the as-of
+                // advances for an existing export, that can only be because the implied capability
+                // has advanced, which means that the write frontier has advanced, which means the
+                // snapshot has definitely been written out.
+                if PartialOrder::less_than(&sink.description.as_of, &frontier) {
```

Comment:
This is fine when `T` is a `TotalOrder`, and maybe we should add that to the `T` bound to make it clear and not complicate the code.

If `T` is partially ordered then `less_than` is not sufficient to conclude that no snapshot is needed because that test returns true if **any** of the timestamps of `description.as_of` `frontier` is greater than the  but we want to check that **all** the timestamps of `frontier` are greater than `description.as_of`. I believe the test for the general `T: Timestamp` case is `description.as_of.all(|t| !frontier.less_equal(t))`, which we also use in reclocking [here](https://github.com/MaterializeInc/materialize/blob/main/src/timely-util/src/reclock.rs#L336).

Reply:
Cool! It looks like adding the bound is not too invasive, so I'll go that route for now.